### ISSUE TITLE
SparseAdam: working with dense parameters but sparse gradients - usecase 

### ISSUE THREAD

**Issue by zabboud** [pytorch/pytorch#101370]
### 📚 The doc issue

I have a use case where some gradients are sparse while others are dense, and therefore I'm splitting them into two groups and optimizing them separately using Adam and AdamSparse. 

In the case of the parameter tensor that is dense while the gradient tensor is sparse (many of the values are set to zero as I do not want to update those parameters), which seems like a perfect use case for the SparseAdam optimizer, however, I get into the situation where if I pass my parameters associated with the sparse gradient matrix (when I say sparse here, I mean that many of the values are zero, not that the gradient tensor is converted to a sparse tensor using `torch.sparse_coo_tensor` or `torch.to_sparse`. 

According to the documentation of the SparseAdam optimizer the variant is applied for the following: 

> In this variant, only moments that appear in the gradient get updated, and only those portions of the gradient get applied to the parameters.

And according to the source code, in the `init()` method, sparse parameters are rejected after being checked with the `param.is_sparse`:
`if sparse_params:
      raise ValueError(
                f"Sparse params at indices {sparse_params}: SparseAdam requires dense parameter tensors"
            )
`
So the optimizer only accepts dense parameters. However, it does not accept dense gradients (which is the purpose of the optimizer). But the problem I am having here, is if I am passing my dense parameters with the sparse gradient tensor (many zeros) the optimizer fails at `optimizer.step` because of this check 

` if not p.grad.is_sparse:
       raise RuntimeError('SparseAdam does not support dense gradients, please consider Adam instead')`

A possible work around is using a hook to change the parameter gradient to a sparseTensor using `torch.sparse_coo_tensor`  but then i get the error 

`Exception has occurred: RuntimeError
hook 'hook' has changed the type of value (was torch.cuda.FloatTensor got torch.cuda.sparse.FloatTensor)`

It is unclear from the documentation how this optimizer is supposed to be used or if this use case can be covered by the sparseAdam optimizer. 

What I would think a possibility of this use-case being covered, which might be useful in many cases where parts of the layers/nodes would be updated, but not others, is passing the parameters and their zero gradients to the sparseAdam optimizer. The optimizer converts the gradient to a sparse tensor to enable this use case. 

If there's a workaround, I'd appreciate the help/input. 

### Suggest a potential alternative/fix

How the sparseAdam optimizer is used - which use cases does it cover, and which use cases it does not. The documentation mentions that it's used for sparse tensors, but not much else. It is specifically confusing because the source code has a check for if the parameter dense is dense (which is required) and the gradient tensor is sparse (which is required), but not on how to satisfy this use-case.

**Comment by janeyx99** [comment]
Thank you for the report! I'm teasing out two particular asks here--please correct me if I am misunderstanding the request:

1. Clarification: when and how should one use the SparseAdam optimizer? Currently, it is true our source code expects dense parameters but sparse gradients, and I understand how that may have friction. cc @cpuhrsch how have other users typically dealt with this? Do they use an optimizer hook, as mentioned by @zabboud? Would it be useful for the optimizer (through a flag perhaps) to automatically convert gradients into sparse tensors?
2. Documentation: Once a recommended way is confirmed, we should document that, with a code sample ideally

**Comment by zabboud** [comment]
@janeyx99 Thank you for your prompt reply. 

To clarify:
1) Yes, the documentation should have when and how one should use the SparseAdam optimizer. Also yes it would be great if there's a flag to automatically convert gradients to sparse tensors. what I would envision this feature to be implemented is this: 

1. User passes dense parameters with dense gradients 
2. Optimizer checks if parameters are dense
3. When running the step(), optimizer converts gradients to sparse tensors (assuming that many of the gradients are set to zero <-- this could be an additional check)

Also thank you for bringing this point up: 

>  Currently, it is true our source code expects dense parameters but sparse gradients, and I understand how that may have friction

That is something I wanted to point out as a possible bug, because based on my understanding having sparse gradients with dense parameters may not be possible? Unless there's a method out there that could do that, then I would like to know how please!


2) Yes, it would be great to have a code sample for how the SparseAdam optimizer is supposed to be used, what specific use cases it currently covers. The current documentation mentions sparse tensors, but is that referring to input tensors? what about sparse parameters or sparse gradients? How do those use cases apply?

Thank you!

**Comment by cpuhrsch** [comment]
I view SparseAdam as an approximation to an algorithm: avoid updating the momentum term if a gradient is zero. This in turn can then be sped up by using sparse layouts.

In general it should be possible to use sparse layouts throughout our entire stack. In practice there might still be enablement gaps.

That means, you shouldn't have to use SparseAdam just to convert some parameters or gradients to a sparse layout at any point within your program.

**Comment by janeyx99** [comment]
To further clarify, SparseAdam is NOT equivalent to normal Adam but for "sparse" gradients/parameters. There is the idea of sparse tensors semantically (like your gradients with lots of 0s) and as a layout (the format represented in torch, like a tensor after running `.to_sparse()`). Our normal optimizers should be able to handle sparse-layout gradients by default (it is worth testing how well they do with sparse parameters though). 

SparseAdam is meant as an _optimization_ for sparse-semantic gradients. The documentation intends to describe the optimization as simply skipping the update computation for the 0-values in a gradient, and the way this happens is through using the sparse layout. This is not to be conflated with the false idea that SparseAdam is meant for sparse layout tensors.

I expect you to still be confused because, in your use case, this seems exactly like what you want--an optimizer that will be optimized for sparse-semantic gradients! And yet, there are still these confusing error messages. The reason for this is the fact that SparseAdam _was indeed_ written for a narrow situation--the situation where you have dense params and sparse gradients. It turns out that there are modules, like `nn.Embedding(sparse=True)`, that output sparse-layout gradients for dense parameters, and this was where SparseAdam could naturally be used. 

We do agree that you should be able to use SparseAdam for your use case as an optimization through some sort of automatic gradient conversion to sparse, but that is not currently available, and I've listed some next steps below. I would recommend you follow your suggestion of using pre-optim hooks to convert the params to sparse in the meantime as further discussion ensues.

- [ ] We should 100% be very clear and document that SparseAdam is meant for a narrow set of dense params with sparse grads and explain when to use the optimizer (as I've explained in the earlier paragraphs). This is actionable and should 100% be done.
-------------------------------------------------
We can then reconsider how optimizers can offer this particular optimization:
- [ ] We should allow dense-layout params with sparse semantics (like your use case) automatically, such as a check that would convert params to sparse-layout tensors back and forth. It is worth noting that we will start off trusting the users' choices--this "optimization" will not be an optimization for tensors without the sparse semantic.
- [ ] We should consider how to offer the best API. Ideally, we would be able to add this optimization as a flag/kwarg to existing optimizers instead of maintaining completely separate instances.

**Comment by zabboud** [comment]
@cpuhrsch  Thank you for your feedback. What you mention is exactly the use-case I am currently working with, but as @janeyx99  mentioned, it is not working for that general use case but a sub-case of that use-case, which is where the `nn.embedding(sparse=True)`. But it is unclear from the documentation that this is the only use case the current implementation covers. 

As @janeyx99 mentioned, the intended use case (and per my understanding) is for sparse-semantic gradients, which should be further improved in future revisions. 

Thank you @janeyx99, more specific wording in the documentation + sample use case would be very helpful to clarify what is currently covered, and possible work around (including the requirement that `nn.embedding(sparse=True)` is required.

I will follow up after having a working work-around for others to reference in the future if they encounter a similar situation. 

Just for my information and others, would the update for having a flag/kwarg for conversion of the gradient tensor to sparse tensor be added as a feature request for future updates of the SparseAdam optimizer?

**Comment by janeyx99** [comment]
Well, using nn.Embedding(sparse=True) is not the only module that falls into the workable subclass, but just an example of one. Don't feel like you need to work that into your model stack somehow. As for a reference workaround--that would be awesome!

Thanks for the feedback on the proposed doc changes.

With regards to the flag/kwarg for converting to sparse--this is something that we haven't designed fully yet. I'm in the middle of revamping torch.optim in general, and so there are lots of pieces to consider (e.g., should we continue fleshing out SparseAdam or should we be adding a generic flag for all optimizers instead?). This feature request (along with the todos under the line) may only get to be considered until months down the line.

Please let me know if you run into trouble with the workaround with using hooks as I believe that should unblock you in the meantime.

**Comment by zabboud** [comment]
@janeyx99 I have attempted to use the hook before calling the optimizer, where I modify the gradient with a pre-hook that both zeros out the desired gradients and modified the gradient to a sparse tensor via this hook function:

`
def exclude_masked(mask):
    def hook(grad):
        modified_grad = grad * mask.float()
        if mask.sum() < mask.numel():
             indices = modified_grad.nonzero().t()
             values = modified_grad[indices[0], indices[1]]
             modified_grad = torch.sparse_coo_tensor(indices, values, modified_grad.size())
        return modified_grad
    return hook
`

The error I get (which I mentioned in my initial post) is the following: 

`hook 'hook' has changed the type of value (was torch.cuda.FloatTensor got torch.cuda.sparse.FloatTensor)`

Is there another way of going around this issue? to satisfy the requirement of having a dense parameter tensor and a sparse gradient tensor? 

thank you

**Comment by janeyx99** [comment]
Ah, I see.

The ideal way to get perf wins from this is to avoid having to process the grads from dense to sparse entirely. If it could be managed (making sparse tensors is not super easy), could you try making your model run in a way where it will just materialize sparse tensors right off the bat? If not, the easier workaround to try is to manipulate your .grad fields right before the optimizer.step().

It came to my brain that the perf wins you may get might not even be worth it anymore if the grads have to be processed anyway and you might just want to use regular Adam. If you have a CUDA machine, fused Adam should be quite performant.

**Comment by janeyx99** [comment]
https://github.com/pytorch/pytorch/pull/101465 should complete the first TODO with documentation.

**Comment by zabboud** [comment]
That works! modifying the gradients right before the optimizer step within the training loop using these few lines did the trick:

```Python
with torch.no_grad():
    for param in parameters_sparse:
          param.grad = param.grad.to_sparse()
optimizer_sparse.step()
```

One question though, technically I should expect some computational gain with using SparseAdam correct? 
 
Yes, I am running on CUDA machines. What would the benefit of fused adam be? Not familiar with it.

**Comment by janeyx99** [comment]
If by computational gain you mean performance win, then yes, but it might be overshadowed by the needing to convert the grads before. Using `Adam(fused=True)` may be more performant--see here for an explanation: https://pytorch.org/docs/stable/optim.html#algorithms

**Comment by zabboud** [comment]
@janeyx99 I just wanted to get more information on the statement you mentioned earlier 

> Well, using nn.Embedding(sparse=True) is not the only module that falls into the workable subclass, but just an example of one. Don't feel like you need to work that into your model stack somehow.

What other scenarios would fall into this workable subclass? where you would naturally have a dense parameter tensor with sparse gradients at instantiation?

**Comment by janeyx99** [comment]
@zabboud If you wanted to use existing torch ops, it looks like the only way is with the nn.Embedding module. If you wrote your own custom op, you could write stuff that would fall under this subclass by having your custom backward output sparse gradients.

**Comment by janeyx99** [comment]
Ah, which optimizer were you using? Would it be SGD? Our foreach ops (the default impl as they are faster) don't currently handle sparse, so you should be able to work around by using:

```
optimizer = SGD(params, ..., foreach=False)
```

I will also submit a more general fallback mechanism when I get a chance.
